### DISCUSSION TITLE
stb_truetype: Glyph Bitmap stride/width issues.

### DISCUSSION THREAD

**Discussion by Ruefren** [nothings/stb#1633] [Q&A] [answered]
Having issues where the stride being the width of the bitmap (calculated by x1 - x0) results in skewed glyphs, but only for some of the glyphs. Changing the stride amount to width + 1 will fix some glyphs but break others that were working. This happens with multiple .ttf and .otf files.

Here is code:

```
i32 x0, x1, y0, y1;
float scale = fontContainer->fontScale; //From stbtt_ScaleForPixelHeight(). pixels parameter is 32.0f
stbtt_GetGlyphBitmapBoxSubpixel(fontContainer->fontInfo, glyphIndex, scale, scale, 0.0f, 0.0f, &x0, &y0, &x1, &y1);
data->w = (x1 - x0); //data->w is int type.
data->h = (y1 - y0);
u32 bitmapLen = data->w * data->h; 
unsigned char* bitmapBuffer = nullptr;
transientStack->Push((void**) &bitmapBuffer, bitmapLen); //Reserves memory on a memory pool for the bitmap
stbtt_MakeGlyphBitmap(fontContainer->fontInfo, bitmapBuffer, data->w, data->h, data->w, scale, scale, glyphIndex);

//Packing the glyphs into an atlas, so this is just the pixel offsets into the atlas.
data->uvX = fontContainer->characterHashTable_Count % 32 * fontContainer->glyphW + 2;
data->uvY = fontContainer->characterHashTable_Count / 32 * fontContainer->glyphH + fontContainer->glyphH + y0;

//Modifying the texture atlas with the new glyph.
Platform_RequestOpenGLContext();
glBindTexture(GL_TEXTURE_2D, fontContainer->glyphMapID);
glTexSubImage2D(GL_TEXTURE_2D, 0, data->uvX, data->uvY, data->w, data->h, GL_RED, GL_UNSIGNED_BYTE, bitmapBuffer);
glBindTexture(GL_TEXTURE_2D, NULL);
Platform_ReleaseOpenGLContext();
```

This is the output of the atlas with the above code:
![image](https://github.com/nothings/stb/assets/101930883/2f0f433f-0596-4048-8500-bf8e91705e43)

If I change the stride parameter of stbtt_MakeGlyphBitmap to data->w + 1 this is the result:
![image](https://github.com/nothings/stb/assets/101930883/e2f8014b-592c-4298-9317-dd5643abbf55)

data->w + 2:
![image](https://github.com/nothings/stb/assets/101930883/78038ea9-4427-430d-9850-6227aa866244)

I also tried using stbtt_GetGlyphBitmap and it had the same output as the first image. That makes me think the issue might be in my code but I'm entirely stumped.

Any ideas?

**Comment by rygorous** [comment, marked as answer]
Try `glPixelStorei(GL_UNPACK_ALIGNMENT, 1)` (haven't checked whether this is the issue in your case, but it usually is.)

**Reply by Ruefren** [reply]
Oh fantastic! This solved my issue, wonder why I haven't run into that before... Thank you!
